clang/docs/checkedc/Update-to-latest-LLVM-sources.md

```diff
@@ -46,6 +53,7 @@ that first.
 You can now branch your baseline branches to create a new master branch:
 
     git checkout -b updated-master
+    git config --local merge.renamelimit 26000
     git merge master
```

Comment:
Next time, when you go from one LLVM release branch to another, you are probably going to have to do an [ours-merge](https://github.com/microsoft/checkedc-clang/issues/947#issue-749005150).  I imagine you don't want to describe the details here since you haven't tested them yet, but perhaps at least link to my comment so you'll see it?

Reply:
Could you please clarify what you mean by an ours-merge? Is this term referring to a git recursive merge with the `ours` option? I'm not sure we would want to do a `git merge recursive ours` for this scenario.

Reply:
The essence of the problem is that the basic principle of an ordinary `git merge` is "take the union of the commits of all the parents".  In a hypothetical future upgrade to LLVM 12, we'd be starting with `updated-master` containing part of the LLVM mainline plus the LLVM 12 release branch up to the tag we use, and `master` containing a shorter part of the LLVM mainline plus the LLVM 11 release branch (up to the tag we used) and the Checked-C-specific commits.  So an ordinary `git merge` will try to retain everything up to our LLVM 12 tag and the Checked-C-specific commits, _plus_ the stuff on the LLVM 11 release branch.  What we really want is to retain everything up to the LLVM 12 tag and the Checked-C-specific commits but _not_ the stuff on the LLVM 11 release branch, which is no longer relevant to the LLVM 12 version we're upgrading to.

If we're lucky and everything on the LLVM 11 release branch is an _exact_ cherry-pick of something later on the mainline, the textual merge will be convergent and the difference will not matter.  But if anything goes wrong, we may get merge conflicts, and even if we resolve any and all merge conflicts, we may be left with some changes specific to the LLVM 11 release branch, so our LLVM baseline is not really equivalent to LLVM 12, and this may cause unforeseen problems later.

The workaround I proposed is to first construct a commit which I'll call X that has the same file content as the LLVM 12 tag we want, but pretend to Git that it also incorporates the portion of the LLVM 11 release branch up to the tag we used.  (This is a lie and may be somewhat confusing when we browse the history later, but it's a well-known technique that is the least evil to handle this scenario given Git's current fundamental limitations.)  We construct X by checking out the LLVM 12 tag and then merging the LLVM 11 tag with `git merge -s ours`.  Then, when we merge X (instead of the LLVM 12 tag) with our previous master, since Git thinks that X already contains the LLVM 11 release branch stuff, it will determine that the way to union everything is to apply the diff from the LLVM 11 tag to X to our master.  Since X has the same file content as the LLVM 12 tag, this will leave us with content equivalent to the LLVM 12 tag plus the Checked-C-specific commits, which is what we want.

Does that all make sense?

Reply:
@mattmccutchen-cci: @kkjeer and I chatted.  Our goal is to follow the clang/LLVM branching strategy.   Our main branch should be based on the clang main branch.  That means our starting commit should have been taken from the clang/LLVM main branch, not a release branch.  Ideally it should be the commit in main used for creating the release branch.  In the past, we sometimes had problems with that exact commit having issues or build breaks, so we may be off by a few commits (clang/LLVM has an optimistic commit strategy).

 If we want to have a version of the Checked C compiler with all patches for a specific release, we would create a release branch of the Checked C clang/LLVM compiler.    It would follow the same pattern that we use overall: start with the actual clang/LLVM release branch, and then merge the Checked C changes into it.  Given that the Checked C clang/LLVM branch should be based off the same commit used for the actual clang/LLVM release branch, the merge should be smooth. There is a possibility that a cherry picked fix for the clang/LLVM release branch conflicts with a Checked C specific change, in which case we'd have to resolve that marge conflict.

Reply:
@dtarditi Sounds OK to me; then this PR should be updated accordingly.  If we do run into stability problems with the LLVM mainline serious enough to get in the way of our development, we can always switch the baseline of the Checked C `master` to an LLVM release branch later; IMO that would be saner than having to do a test merge of a Checked C branch with the LLVM release branch every time we want to validate a Checked C change.  Or if we can identify a small number of commits on the LLVM release branch that fix the problem affecting us, we could cherry-pick them and then revert the cherry-picks at the beginning of the next baseline upgrade process.

It would be nice if you had responded [here](https://github.com/microsoft/checkedc-clang/issues/947#issuecomment-736844803) before I took the time to write that wall of text, though I know you're busy.  In any event, now you know what it would take if you did want to use successive LLVM release branches as baselines for Checked C.